packages/components/src/border-control/styles.ts

```diff
@@ -0,0 +1,177 @@
+/**
+ * External dependencies
+ */
+import { css } from '@emotion/react';
+import type { CSSProperties } from 'react';
+
+/**
+ * Internal dependencies
+ */
+import { COLORS, CONFIG, rtl } from '../utils';
+import { space } from '../ui/utils/space';
+import {
+	StyledField,
+	StyledLabel,
+} from '../base-control/styles/base-control-styles';
+import { BackdropUI } from '../input-control/styles/input-control-styles';
+import { Root as UnitControlWrapper } from '../unit-control/styles/unit-control-styles';
+
+import type { Border } from './types';
+
+const labelStyles = css`
+	font-weight: 500;
+`;
+
+export const BorderControl = css`
+	position: relative;
+`;
+
+export const InnerWrapper = css`
+	border: ${ CONFIG.borderWidth } solid ${ COLORS.gray[ 200 ] };
+	border-radius: 2px;
+	flex: 1 0 40%;
+
+	/*
+	 * Needs more thought. Aim is to prevent the border for BorderBoxControl
+	 * showing through the control. Likely needs to take into account
+	 * light/dark themes etc.
+	 */
+	background: #fff;
+
+	/*
+	 * Forces the width control to fill available space given UnitControl
+	 * passes its className directly through to the input.
+	 */
+	${ UnitControlWrapper } {
+		flex: 1;
+		${ rtl( { marginLeft: 0 } )() }
+	}
+`;
+
+export const WrapperWidth = ( width: CSSProperties[ 'width' ] ) => {
+	return css`
+		width: ${ width };
+		flex: 0 0 auto;
+	`;
+};
+
+export const BorderControlDropdown = css`
+	background: #fff;
+	${ rtl(
+		{
+			borderRadius: `1px 0 0 1px`,
+			borderRight: `${ CONFIG.borderWidth } solid ${ COLORS.gray[ 200 ] }`,
+		},
+		{
+			borderRadius: `0 1px 1px 0`,
+			borderLeft: `${ CONFIG.borderWidth } solid ${ COLORS.gray[ 200 ] }`,
+		}
```

Comment:
Looking at the [implementation of `rtl`](https://github.com/WordPress/gutenberg/blob/8ef2008d3c90b101d11ade7f422f6517867411a1/packages/components/src/utils/rtl.js), we may have to "break" the `borderRadius` so that we have explicit "left" and "right" props

Reply:
I was under the impression that the first argument to [`rtl`](https://github.com/WordPress/gutenberg/blob/8ef2008d3c90b101d11ade7f422f6517867411a1/packages/components/src/utils/rtl.js#L63-L81) were the styles for left-to-right and if supplied the second argument were the desired styles for right-to-left. The left-to-right styles were only automatically converted if the right-to-left weren't provided.

I'm wondering if the memoization of the style classes is causing a problem? It's highly likely I'm missing something around the use of `useCX` as well.

I've tried adding `rtl.watch()` to the `useMemo` dependencies within the component hooks as per `rtl.js`'s inline comments. That still didn't seem to make the styles update reliably in the Storybook. After a fresh build with RTL already toggled on, the correct styles are generated for me as per screenshot below. That would suggest to me that the highlighted call to `rtl` is actually ok.

<img width="297" alt="Screen Shot 2022-03-16 at 3 58 49 pm" src="https://user-images.githubusercontent.com/60436221/158526606-86e626e4-68f2-4933-ada7-fc862b50142a.png">

Toggling LTR and RTL from here still doesn't reliably update the styles. Any ideas?

Reply:
> I was under the impression that the first argument to [rtl](https://github.com/WordPress/gutenberg/blob/8ef2008d3c90b101d11ade7f422f6517867411a1/packages/components/src/utils/rtl.js#L63-L81) were the styles for left-to-right and if supplied the second argument were the desired styles for right-to-left

That's actually correct, I must have somehow misread the code at the time of my first comment in this convo.

> Toggling LTR and RTL from here still doesn't reliably update the styles. Any ideas?

I can confirm that all RTL styles are currently not being updated correctly.

The reason seems to be because the calls to `rtl` are not the result of a function return (instead, they are just part of a `SerializedStyle` that gets calculated once and never again).

To solve this issue, it should be enough to change those values holding a `SerializedStyle` to a function returning a `SerializedStyle` — this allows this function to run again when needed, and return the correct RTL styles.

<details>
<summary>click to expand</summary>

```diff
diff --git a/packages/components/src/border-control/border-control-dropdown/hook.ts b/packages/components/src/border-control/border-control-dropdown/hook.ts
index f180cec6fb..d2d32e6b20 100644
--- a/packages/components/src/border-control/border-control-dropdown/hook.ts
+++ b/packages/components/src/border-control/border-control-dropdown/hook.ts
@@ -52,7 +52,7 @@ export function useBorderControlDropdown(
 	// Generate class names.
 	const cx = useCx();
 	const classes = useMemo( () => {
-		return cx( styles.BorderControlDropdown, className );
+		return cx( styles.BorderControlDropdown(), className );
 	}, [ className, cx ] );
 
 	const indicatorClassName = useMemo( () => {
diff --git a/packages/components/src/border-control/border-control/hook.ts b/packages/components/src/border-control/border-control/hook.ts
index af54a18051..e242da33c5 100644
--- a/packages/components/src/border-control/border-control/hook.ts
+++ b/packages/components/src/border-control/border-control/hook.ts
@@ -117,15 +117,15 @@ export function useBorderControl(
 		const widthStyle =
 			!! wrapperWidth && styles.WrapperWidth( wrapperWidth );
 
-		return cx( styles.InnerWrapper, widthStyle );
+		return cx( styles.InnerWrapper(), widthStyle );
 	}, [ isCompact, width, cx ] );
 
 	const widthControlClassName = useMemo( () => {
-		return cx( styles.BorderWidthControl );
+		return cx( styles.BorderWidthControl() );
 	}, [ cx ] );
 
 	const sliderClassName = useMemo( () => {
-		return cx( styles.BorderSlider );
+		return cx( styles.BorderSlider() );
 	}, [ cx ] );
 
 	return {
diff --git a/packages/components/src/border-control/styles.ts b/packages/components/src/border-control/styles.ts
index 624d98425d..a52a46d571 100644
--- a/packages/components/src/border-control/styles.ts
+++ b/packages/components/src/border-control/styles.ts
@@ -26,7 +26,7 @@ export const BorderControl = css`
 	position: relative;
 `;
 
-export const InnerWrapper = css`
+export const InnerWrapper = () => css`
 	border: ${ CONFIG.borderWidth } solid ${ COLORS.gray[ 200 ] };
 	border-radius: 2px;
 	flex: 1 0 40%;
@@ -55,7 +55,7 @@ export const WrapperWidth = ( width: CSSProperties[ 'width' ] ) => {
 	`;
 };
 
-export const BorderControlDropdown = css`
+export const BorderControlDropdown = () => css`
 	background: #fff;
 	${ rtl(
 		{
@@ -150,7 +150,7 @@ export const ResetButton = css`
 	}
 `;
 
-export const BorderWidthControl = css`
+export const BorderWidthControl = () => css`
 	/* Target the InputControl's backdrop */
 	&&& ${ BackdropUI } {
 		border: none;
@@ -178,7 +178,7 @@ export const BorderStyleButton = css`
 	}
 `;
 
-export const BorderSlider = css`
+export const BorderSlider = () => css`
 	flex: 1 1 60%;
 	${ rtl( { marginRight: space( 3 ) } )() }
```

One more note — because these static (or "dynamic") `SerializedStyles` are not a React component, we should probably rename them to start with lowercase. Only Styled components should start with a capital letter

```js
// `SerializedStyles`: calemCase
const borderControlDropdown = css` /** styles */ `;

// styled components: PascalCase
const InnerInput = styled.input` /** styles */ `;

Reply:
Thanks for the help sorting out the RTL styling. 🙇 

I've pushed these fixes and the naming corrections. I think this should be getting close now 🤞 .

Reply:
Thank you! And sorry if I took so long to answer back

Reply:
No worries here, these things move at their own pace. 

I appreciate the help in improving the quality of this component. I'm still learning the best practices around Emotion and our components package. So this is all valuable experience for me.